cclib/parser/logfilewrapper.py

```diff
@@ -0,0 +1,312 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2023, the cclib development team
+#
+# This file is part of cclib (http://cclib.github.io) and is distributed under
+# the terms of the BSD 3-Clause License.
+
+# TOOD: This file belongs in cclib.io, but circular dependency issues mean it can't go there just now.
```

Comment:
What's the circular dependency?

Reply:
Isn't that the question :)

1) `cclib.parser.logfileparser` imports `FileWrapper` (currently from `cclib.parser.logfilewrapper`) in order to be backwards compatible (so you can pass plain strings to `Logfile`'s constructor and it will automatically wrap with `FileWrapper`)
2) `cclib.io.ccio` imports all parsers which inherit from `cclib.parser.logfileparser.LogFile` in order to return the program specific parser it needs
3) `cclib.io.__init__` imports `cclib.io.ccio`, purely for convenience

Thus `cclib.parser.logfileparser` is unable to import anything under `cclib.io`, because the `__init__` file loads `cclib.io.ccio`, which in turn imports `cclib.parser.logfileparser` again.

I think this should be fixable if `cclib.io.__init__` imports `FileWrapper` early (which thankfully has no cclib dependencies), then you could do `from cclib.io import FileWrapper` and that should avoid the circular dependency I think (although doing `import cclib.io.logfilewrapper` would still be problematic). However, `cclib.io.cjsonreader` also import stuff back from `cclib.parser`, complicating matters. I suspect a lot of bits and pieces could do with moving around here, so maybe it's best for this to wait until a v2.x is in the works?

Reply:
Breaking the dependency can wait.  The big problem is number 2.  I think there is a design pattern that helps with this particular case.